envoy/matcher/matcher.h

```diff
@@ -25,7 +26,7 @@ class ServerFactoryContext;
 namespace Matcher {
 
 using MatchingDataType = absl::variant<absl::monostate, std::string>;
-
+const char DefaultMatchingDataType[] = "string";
```

Comment:
If we think this type name should be pre-defined or pre-declared, may be `constexpr absl::string_view` would be better.

And the `supportedDataInputTypes` could return the set of string view. The `dataInputType`  could also return a  string view.

For a big matcher, this way could reduce lots of unnecessary string copy or memeory allocation in the config loading phase. (With the more and more complex configuration, I think sometimes we should also concern the performance at the config/xds loading)

Reply:
IMHO, the string_view is one of common over-used concepts, especially for function return type and stored in a container or class member, which introduces unnecessary concern about lifetime but doesn’t save anything. Because 1)in most cases of std::string, there are actually no copies at all and 2)In the case when the copy of std::string is required because the lifetime of object must be guaranteed, string_view doesn't save anything since it requires the copy when it is converted to std::string() 

For the definition here, there is no difference between `const char[]` and `string_view`. string_view is just like char*, a pointer with length.
For copy, like the example below, there is no copy at all c++17 onwards, In earlier language modes, this is usually RVO (go/totw/11) and is guaranteed to be at worst a move. ```
auto input_type = data_input_->dataInputType();
```

In summary, `string_view` is usually suggested for function arguments (probably the only place safe to use without careful consideration.) but not encouraged in containers or class members. Storing string_view is source of programmer error. More details can be found https://abseil.io/tips/180. Also here https://abseil.io/tips/101 about copies.

Reply:
absl::string_view provides you with a stronger type safety and optimizations. It is almost always preferable over `char[]` or `char*`. For example, down below you have this code:

```
 auto input_type = data_input_->dataInputType();
 if (input_type != DefaultMatchingDataType) {
```

what would  happen with the existing code like above if the dataInputType() method is changed line this?

```
 virtual const char* dataInputType() const
```

Reply:
Re @yanavlasov  I don't think people should change `dataInputType` to `const char*` in the first place. IMHO, that is a hypothetical assumption that probably won't and should not happen.  I am neutral on this though as it is just simple const declaration.

Re @wbocope : I don't think we should use `string_view` in the container as it is source of programmer error. In terms of performance, I think benchmark to prove the performance overhead is needed. We should not do micro-optimization the code (just for 6 chars "string") without real benchmark data while introducing the bad programming principle. Also, as i mentioned before we should not hardcode/pre-declared for custom matcher. I think demangle typeid is better approach. So, I am not convinced to change to string_view.

Reply:
And you can do a quick search to the code tree, there should lots of method to return string view and also lots of method to return container with string view.

the lifetime problem is important. But in this scenario, for the static string to represent a type, we need worry it.

Reply:
I don't think those `return string view and also lots of method to return container with string view` means they are good example. In contrast, IMHO I think we should follow general prince https://abseil.io/tips/180 which make more sense.

Reply:
Yeah, it is good tips. And should be obeyed always.

But I don't know why return string view violate this rule. esp we know exactly there will no dangling reference.

Basically string view is a reference, and obviously returning references is a very common practice.

The key point is no dangling reference, not no reference.

Reply:
Return string view doesn't violate this. But from my past experience , many places on the caller site you will still need to convert string_view to string to extend its life time.
 
Returning reference doesn't save things as you may expect but introduce lifetime concern in general. Nowadays compiler is smart to optimize away many copies if you just return by value. https://abseil.io/tips/101  We always need extra care about reference like pointer.

IMHO, robust/easy programming is also encouraged and important in C++. Same philosophy like raw pointer should be more performant than smart pointer but it is not safe to use and not encouraged in modern c++ nowadays.

Reply:
So, it depends. In some scenarios string should be returned.

And in this scenario, no lifetime problem, no need extend the lifetime, it's only be used for one time searching after it be returned. This scenario shoud be the case where string view could be used.

Smart pointer is different with string. Smart pointer is cheap, unique pointer is especially cheap. But string is expensive.

Reply:
Anyway, seems there's tech disagreement.

But it doesn't matter. I will mark this as unresolved and let other maintainers to make the decision.

Except this point, others lgtm. Thanks for your contributions.

Reply:
I don't think string is cheaper than smart pointer or cost of them two are comparable. string has small string optimization. There is no doubt that smart pointer is more expensive than raw pointer as it  involves extra overhead (like the cost of creating and managing the unique pointer object itself or ref_count of shared_pontter).

My smart pointer is just an example to prove why the robust and simple programming principle is desired.  People should not spend time to figuring out whether it is safe or not for very specific when there is no benchmark data proves the performance gain or gain is negligible.

Reply:
As long as we don't store the returned string_view of function, for the case here that is `inline constexpr absl::string_view` (to ensure safe const in header), string_view will force a copy to string on the caller site to ensure the lifetime when needed. So, i am ok with either returning string or string_view in this particular case though i don't think returning string_view/reference is encouraged way in general as described above.

Reply:
Typically, if there is no special liftetime problem, the object should return const string& or string view to avoid unnecessary new string creation.
The object should ensure the view/ref has liftetime no shorter than the object self.
**This is common practice rather than specially optimization.**

(RVO is helpful for case where we intend to create new string/object, but not enough if we can complete avoid the creation)

Reply:
> The object should ensure the view/ref has liftetime no shorter than the object self. This is common practice rather than specially optimization.

I don't think the view/ref's lifetime can be guaranteed in C++. Otherwise, raw pointer should be always used (It should be used when there is no ownership transfer ). So same to raw pointer, every time people see the ref, they will need to think a bit more about the lifetime of the object it refers to. Yes, the static/global objects lifetime is guaranteed but it should be very few of these object in production code. An common use case is to return ref of class member in class method and this is because this lifetime of method is guaranteed to same as class object. Unless the guarantees like example above, I don't think return ref is encouraged.  Very common mistake of new programer is returning the ref of local variable.

I agree the performance optimization is good but making code more complicated when there is no performance gain in first place is not good tradeoff. Not just RVO. For example `Returning std::string, initializing std::string` In C++17 onwards, no copy or move is performed . In earlier language modes, this is usually RVO and is guaranteed to be at worst a move.

Reply:
>  Another common use case is to return ref of class member in class method and this is because this lifetime of method is guaranteed to same as class object.

I think I mainly mean this scenario. Return the ref of class member, ref of parent class member, global object, etc.

I agree we shouldn't make the code more complicated for uncertain performance improvement.
But typically we cannot agree on what is complicated code for now.

We do too much disscussions here about this topic, just let's move to the next.